contracts/contracts/token/OUSD.sol

```diff
@@ -584,20 +633,111 @@ contract OUSD is Initializable, InitializableERC20Detailed, Governable {
             ? MAX_SUPPLY
             : _newTotalSupply;
 
-        _rebasingCreditsPerToken = _rebasingCredits.divPrecisely(
-            _totalSupply.sub(nonRebasingSupply)
-        );
+        _rebasingCreditsPerToken =
+            (_rebasingCredits * 1e18) /
+            (_totalSupply - nonRebasingSupply);
 
         require(_rebasingCreditsPerToken > 0, "Invalid change in supply");
 
-        _totalSupply = _rebasingCredits
-            .divPrecisely(_rebasingCreditsPerToken)
-            .add(nonRebasingSupply);
+        _totalSupply =
+            ((_rebasingCredits * 1e18) / _rebasingCreditsPerToken) +
+            nonRebasingSupply;
 
         emit TotalSupplyUpdatedHighres(
             _totalSupply,
             _rebasingCredits,
             _rebasingCreditsPerToken
         );
     }
+
+    function delegateYield(address from, address to)
+        external
+        onlyGovernor
+        nonReentrant
+    {
+        require(from != to, "Cannot delegate to self");
+        require(
+            yieldFrom[to] == address(0) &&
+                yieldTo[to] == address(0) &&
+                yieldFrom[from] == address(0) &&
+                yieldTo[from] == address(0),
+            "Blocked by existing yield delegation"
+        );
+        RebaseOptions stateFrom = rebaseState[from];
+        RebaseOptions stateTo = rebaseState[to];
+
+        require(
+            stateFrom == RebaseOptions.NotSet ||
+                stateFrom == RebaseOptions.StdNonRebasing ||
+                stateFrom == RebaseOptions.StdRebasing,
+            "Invalid rebaseState from"
+        );
+
+        require(
+            stateTo == RebaseOptions.NotSet ||
+                stateTo == RebaseOptions.StdNonRebasing ||
+                stateTo == RebaseOptions.StdRebasing,
+            "Invalid rebaseState to"
+        );
+
+        if (
```

Comment:
This is a good point. As an extra defensive check I altered the `isNonRebasingAccount` to never return true when accounts are part of yield delegation. While arguably we would only need to check the yield delegation sources, I think it is a good code readability measure to include both yield delegation rebase types.

Reply:
I think there's a bit of confusion happening because the function `_isNonRebasing()` now has two possible non-overlapping meanings. Does it mean the account does not rebase, and the function is about the accounting on the account? Does it mean that the account is a StdNonRebasing account, and the function is about account type? In the past, these were equivalent, but now those are different questions with potentially different answers on the same account.

Probably time to make this clear in the code.

### delegateYield()

Let's start with `yieldDelegation`. What we really need here for the from/to addresses is to:

```
if  already good account type / accounting
  allow
else if account type / accounting can be migrated:
  migrate
else
   reject
```

But there's a duplication in separating good accounts from good accounts that need migration, so I think this is equivalent, and a little simpler to write

```
require(usable account types)
if needs migration
  migrate
```

For the check if it needs migration, we want to look at the accounting (`alternativeCreditsPerToken`), rather than account types. We also don't want a second automatic behind the scenes migration like the current `_isNonRebasing()` does. Instead, we should just directly compare `alternativeCreditsPerToken` to zero when deciding to migrate or not.

### _rebaseOptIn()

This should follow a similar pattern as yield delegation, rather than calling `_isNonRebasingAccount()`. It should check that the account is of the allowed type, and then that the accounting needs to be migrated. This is what _rebaseOptOut already does.

### _adjustAccount()

At this point _adjustAccount would be the only remaining code using `_isNonRebasingAccount()`. Let me do some more thinking on that.

Reply:
### _adjustAccount() continued

What if we changed `_isNonRebasingAccount()` to `_autoMigrate()` and only kept the first if in the code.

Then we could change `_adjustAccount()` to something like

```
if yieldSource
  ...
  return
else if yieldTarget
  ...
  return
  
_automigrate(account)
if alternativeCreditsPerToken >0
  ...
else
 ...

Reply:
I think this has cleared up rather nicely: origin/sparrowDom/rebaseElsewhere_v2

Reply:
Yes, way nicer.

Do we want the automigrate in rebaseOpt  `_autoMigrate(_account)`? I'm don't necessarily think it's wrong, I'm just wondering about the pro's and cons.

Reply:
good point we don't actually need the auto migration we can just explicitly allow for the empty contracts to pass the rebaseOptIn: https://github.com/OriginProtocol/origin-dollar/pull/2298/commits/d7a303f14e125b047935cf9e2e64554b4e382598